### ISSUE TITLE
Should CNRs be mandatory?

### ISSUE THREAD

**Issue by mkurzydlowski** [erasmus-without-paper/ewp-specs-architecture#30]
Some developers suggested making CNRs mandatory.

Currently, both implementing them as a server (they are defined as separate APIs) and as a client (they are marked by fields in the APIs) is optional. Partners can currently add support for CNRs (both client and server side) as they wish. They can start with a minimal set of APIs and implement other as they progress.

One of the sections in the architecture overview describes the concept of CNRs:

https://github.com/erasmus-without-paper/ewp-specs-architecture#broadcasting-changes-cnr-apis

It should be added that currently many partners issue CNRs manually. SIS users send a change notification by pushing a button. They may do it when they wish or they might never do it in a specific case.

As long as there is no technical and business guarantee that a change notification reaches the partner it seems that there is no way one can rely on CNR and that there is no way around periodic or manual ["syncing" of local data](https://github.com/erasmus-without-paper/ewp-specs-architecture#index-pulling-and-modified_since).

**Comment by fmapeixoto** [comment]
Let me start from the beginning so I can express myself properly and with no confusions.

When EWP was created the main purpose (and I think everyone agrees) was to lower the paperwork of IROs and, in addition to that (correct me if I am wrong), lower the interactions needed between IRO offices and with the systems.

I totally agree that EWP should be modular and HEIs do not have to implement all APIs at once but rather go as they pleased and one by one until it fits their purpose (and legal obligations).

But I also believe that some APIs should be implemented in parallel as they complement each other. As Institutions API, OUnits API, can exist separately, I believe IIAs APIs, for example, should be in an implementation group, if you want them to fit the whole purpose of correct IIA exchange.
Does it make sense to have Get without Index in IIAs? Can you exchange and sign agreements without the approval API?

Given this, the CNRs were built to lower the e-mails communications between users and if I send a CNR I don't have to send an email, and this is the main purpose of the EWP network, right?
CNRs have defines responses that you can rely, verifying  that a message has arrived, am I correct?
How can you trust that an e-mail has arrived? If you don't receive a reply you have to call the services, or send another e-mail, etc. I am not saying CNRs can't fail, not at all, I am saying that they must be the primary way of communication and, if it fails, like any other system does, then you should have another method to retrieve the information, of course.

For me, the whole IIA APIs, if an HEI want to exchange IIAs within the network, should be mandatory. This guarantees that the whole digitalisation process is covered from the beginning.

**Comment by kkaraogl** [comment]
As a member of the development team behind the OLA application, I would also be in favor of a somewhat push notifications mechanism for exchanging changes in Learning Agreements via the EWP. I am afraid that trying to exchange LA changes without it, could potentially have negative effects in the application's load and general well-being, long-term.

**Comment by dimidasko** [comment]
A comment on:
> As long as there is **no technical and business guarantee** that a change notification reaches the partner it seems that there is no way one can rely on CNR and that there is no way around periodic or manual "syncing" of local data.

The "technical" quarantee that a CN reaches the partner is by the current design, which requires the receiver of the CNR to confirm reception with a new message to the sender. Otherwise the sender will have to maintain the CN as pending for ever. We may make improvements by requiring resending of pending CNs, until final acknowledgment.

The "business" quarantee comes from accepting a common spec that all partner nodes have the responsibility to honor. If we make CNRs mandatory, responsibility lies with the partner for implementing CNRs correclty and sending/receiving correct information. We as network operators do not have to check how the partner implements them. It's in their own interest to do it right and according to spec.

The argument in favor of CNRs is that it will ease the load on each consumer to check and recheck local data for possible changes with the "modified since" mechanism. And while this may be feasible for one node representing a couple of institutions, it will be much harder for a central node, such as the dashboard, which needs to proxy hundreds of institutions and maintain all cached data in sync with senders. Without CNRs, network traffic and node load to check validity of local data will be unnecessarily high.  

The argument against (and my concern) is the difficulty of implementing the CNRs at each EWP node, where all the current implementations will have to integrate a publish-subscribe mechanism for each and every record change in our databases. Ideally, to solve this problem we could adopt a message broker open source solution (e.g. RabbitMQ) that implements this mechanism without us all having to worry about maintaining states for pending CNs to be acknowledged by consumers. But I am not sure how this could be integrated with the current EWP communication protocols in place.

In my view, it's a trade off between resource wasting processes and complexity of implementation. But if the implementers of the centralized solution feel that CNRs are necessary for lighter operation of their proxy, their say should weigh in heavily.

**Comment by BavoNootaert** [comment]
Where in the specification is there a message from the CN receiver to the sender? Or do you mean the HTTP response? When the response contains a 2xx response, the sender can assume that the message was delivered. If the receiver later detects that an error occurred (either during storing the notification or processing it), it is the responsibility of the receiver to query the index and process all the changes it has missed.

I don't see a need to use a message broker. An implementation can be as easy as a field in the database which the user can use to find whatever still needs to be sent to the partner, what was sent, what generated an error, etc...

I agree that if you frequently need to query lots of others hosts to obtain the current state, that will be too expensive. This is not limited to the dashboard, but applies to all hosts.

**Comment by georgschermann** [comment]
We are currently maintaining dozens of hosts with hundreds of heis, the load of repeatedly querying all other hosts is significant and will grow in time. On the other hand relying on the CNRs 100% is also troubling because of various technical and business case reasons. I think the best case scenario would be to strongly recommend CNR API for small providers, maybe make it mandatory for larger providers with lots of universities and also require the possibility of a manual, user-triggered possibility to query for changes.

**Comment by jpbacelar** [comment]
It seems clear from the discussion above that the volume of data exchanges EWP is expected to handle, and the fact that a lot of such traffic will be concentrated in a few nodes, makes mandatory CNRs, duly complemented by a fall-back solution, a logical evolution of the current spec. 

Systems such as the OLA are built in such a way that users receive real time notifications of the actions relevant to them. Relying on scanning the network for updates overnight would amount to a rustic step-down; alternatively scanning it every 10 or 15 minutes would entail an exponential (and extremely inelegant) growth of perfectly avoidable requests. 

Making CNRs mandatory also seems essential to future-proof the network and pave the way for new use cases. For example, the new mobile App will consume a lot of IIA information, which needs to be put at the students fingertips in an efficient and scalable way; we don’t usually discuss how our actions can improve the student experience, but that’s actually one of the key concerns to bear in mind when making seemingly technical decisions. 

Lastly Francisco’s point about IROs risking to have to rely on email even when they have a network connection is a powerful illustration of what we can (easily) do better. As long as we are not taking full advantage of the infrastructure we built we will fail to deliver the kind of upgrade of the IRO flows that set the project in motion in the first place. 

Seeing how this is a requirement put forward by the reference implementation (which handles more than half of the network users) as well as an issue where there is agreement among a majority of the project technical partners I think it’s only right to move forward and share some good news with our community soon, perhaps starting with the APIs that are currently being updated.

**Comment by megachirone** [comment]
I totally agree with the ones saying that CNRs should be mandatory. In our implementation we strongly rely on them to securely let the partner know whenever there are new data to be acquired and viceversa!

**Comment by CarmenDiazSIGMA** [comment]
We agree with idea that CNRs should be mandatory. 
We think that this is the best help for IROS, since this allows them not to be constantly checking if there have been changes, which minimizes the volume of validations that they have to do.
When we've explained to the universities how to use our integrated solution with EWP, they all have the same concern, how do we know if the partner has made changes and can we obtain the information? Will it be necessary to consult institution to institution to know the updated status of the information? 
In our opinion EWP must address these concerns as the project was created with the goal of improving the current process.

**Comment by umesh-qs** [comment]
Even though CNR is not full proof, they still should be made mandatory to cover the majority of the cases.
For such cases where CNR doesn't work, our clients have the option to refresh specific partners data manually. 
That way we distribute the load, rather than running constant sync or nightly sync.

**Comment by Emanuele-Fortunati** [comment]
I write in behave of AEC (host for 90+ HEIs) and I too believe the CNR should be mandatory and implemented by the hosts as soon as they start implementing the APIs that need/support the use of it.
I think if the aim of EWP is to reduce paper(work), then we should aim to create reliable systems, like the CNR, which fulfil the same purpose. The benefit of an automated system (if well made) is also to reduce the possibility of human errors, which, eventually reduce the paperwork even more.
Of course, notification systems like the CNR can be very expensive in terms of server usage, especially if they have been built on top of existing APIs or systems which don't naturally support such notification scenarios. Making them mandatory will require providers to take that into account, but I believe that the benefits in time will surely compensate for the server and infrastructure costs.

**Comment by tsiakmaki** [comment]
I also agree with the idea of making CNRs mandatory. It is going to save all hosts from unnecessary load, as it will reduce the need of checking possible changes that took place at other hosts. 

A current alternative in case of a failure is the index endpoint. I would be in favor of characterizing the 'modified_since' parameter as *MUST* at the [documentation](https://github.com/erasmus-without-paper/ewp-specs-api-omobility-las/blob/master/endpoints/index.md#modified_since-optional), for the same reason.

**Comment by sascoms** [comment]
We also strongly agree and think that CNRs should be made mandatory.

As a quick example: 
One of our clients has ~400 partners. ~800 IIAs. 
If CNRs are not implemented on those partner systems, the IRO staff need to check each partner's IIA index manually one by one.
Or we as providers need to automate this checking updates (even if there are no remote changes) regularly with some nightly/daily sync jobs.

And automated jobs for checking changes will cause a gradually increasing huge unnecessary traffic.

Why create this unnecessary traffic if there are no remote changes? Or if the CNRs are implemented by all and works well, would this unnecessary traffic exist?

Just a simple/sample stat: Only one of our HEIs now have ~90k incoming requests only in 2 months. 


Another point, we also think that CNRs shall not be considered separate from the APIs.
When an API is being implemented, sending/receiving CNRs must be implemented too. And the registry API test-suite should test the related CNRs together.
If the CNR of the API is not ready or has errors, then the API shall not be permitted to be activated.


Additionally as a suggestion:
When we send a CNR to a partner HEI, we accept the 200 header as it has been received and no problem occurred.
However, it would be better to receive a response with success/failed (true/false + msg) information instead of only relying on the header codes.

In our CNR implementations between our EWP gateway and the client IRO system, all our CNRs have a result object returned from the IRO system to our EWP gateway.
This way, our EWP gateway knows if the conveyed CNR is received and saved by the HEI system. If not, it returns the appropriate 40x header responses to the sender HEI.

Also, we have both automatic and manual CNRs. CNRs can also be manually fired by the IRO users. 
IRO staff can list and see successfully sent CNRs as well as the CNRs failed to be sent.

And in our EWP gateway software, we have created a monitoring application that logs all request failures so that we can catch the network or system errors and fix them in order just to make the communication go on smoothly.


One last point, we think that the decisions and principles agreed should apply to all of the providers, either small or large, not only to a subset.


To sum up, making CNRs mandatory and implemented by all, with a neat response return, will save resources and contribute better to the paperless smooth communication between HEIs/IROs.

**Comment by j-be** [comment]
Through the discussion on https://github.com/erasmus-without-paper/ewp-specs-api-omobility-la-cnr/issues/2 I was made aware of this thread. My thoughts:

My first critique point is, that everybody seems to talk about **receiving** a CNR through an API endpoint (i.e. the thing you `POST` to), but the spec as I understand it fails to mandate that CNRs are actually **sent** (i.e. doesn't require me to actually `POST` to the endpoint if I have changes). So, just accepting CNRs by sending back a hardcoded `200` and the appropriate XML, but never actually sending CNRs for changes happening at my side would be well within spec as I understand it. I agree it is a somewhat counterintuitive interpretation, but from my point of view it would be a legal one, e.g. it always says `wants to broadcast a notification`.

EDIT: I just now noticed the 2 commits to erasmus-without-paper/ewp-specs-api-omobility-las and erasmus-without-paper/ewp-specs-api-iias - my bad. So it seems that pretty much tackles my critique. The only thing I still don't like is the fact, that main architecture spec (i.e. this repo) does not already clarify, that CNR sending is required, too. Below is still relevant though.

On a different topic, it reads like CNRs are regarded by some to be the silver bullets for keeping caches coherent. But that is in contradiction with the statement `you MUST NOT assume that implementing CNR APIs is sufficient to keep your copy of the data up-to-date`, right?

I guess there are still HEIs (we would be one of them), which opt for having a *we don't store anything and fetch everything on demand* approach. Switching from there to *we additionally cache everything for 24 hours* is easy enough, and should deliver good performance. And from my point of view for things like LA or IIA a latency of 24 hours is more than acceptable - I don't think current implementations relying on sending actual paper around can even match that. For those handful of times, where it is super urgent, one can still pick up the phone, and the other side may click a "Refresh" button. In such a scenario, CNRs do not serve any meaningful purpose.

Following that thought, CNRs are sent out on **every** change, not on every **relevant** change. I.e. if I where to fix a typo in an LA's `sending-hei.ounit-name` there is no need for receiving LA to do anything. Yet, in order to be able to determine whether a CNR I receive requires reaction from my side, I would need to hold a complete persistent copy of everything. Else I can't even find out what changed, not to mention if I need to notify anybody about the change. That though is in contradiction with `Avoid storing the data permanently`.

I think that limits usefulness of CNRs for anybody not trying to implement coherent caching. So, I would propose to add some `reaction-requested` flag, to CNR where either spec or the implementer can express, whether it thinks a user at receiving HEI should be notified as manual interaction is required to move on (i.e. receiving HEI needs to sign a LA's `changes-proposal`). So those not trying to implement coherent caching can still send e.g. Notification-EMails around, and get at least some kind of feature out of having to implement CNRs.
